hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncRequestFutureImpl.java

```diff
@@ -543,22 +572,36 @@ void sendMultiAction(Map<ServerName, MultiAction> actionsByServer,
             && numAttempt % HConstants.DEFAULT_HBASE_CLIENT_RETRIES_NUMBER != 0) {
           runnable.run();
         } else {
-          try {
-            pool.submit(runnable);
-          } catch (Throwable t) {
-            if (t instanceof RejectedExecutionException) {
-              // This should never happen. But as the pool is provided by the end user,
-              // let's secure this a little.
-              LOG.warn("id=" + asyncProcess.id + ", task rejected by pool. Unexpected." +
-                  " Server=" + server.getServerName(), t);
-            } else {
-              // see #HBASE-14359 for more details
-              LOG.warn("Caught unexpected exception/error: ", t);
+          boolean completed = false;
+          int nbTry = 0;
+          while(!completed) {
+            try {
+              ++nbTry;
+              pool.submit(runnable);
+              completed = true;
+            } catch (Throwable t) {
+              if (t instanceof RejectedExecutionException) {
```

Comment:
IIRC, for simplify the logic, we use LinkedBlockingQueue without a size limit for the thread pool there so we will not get the RejectExecutionException by default, unless users specify a thread pool manually?

Reply:
At least in my environment:

```
2019-04-02 12:04:47,726 WARN  [Executor task launch worker for task 23] AsyncProcess:1004  - #8, the task was rejected by the pool. This is unexpected. Server is XXXXXX.fr.world.socgen,16020,1552142849868
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@6eb15d4f rejected from java.util.concurrent.ThreadPoolExecutor@73168fb9[Running, pool size = 128, active threads = 128, queued tasks = 0, completed tasks = 182]
            at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
            at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
            at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369 
```

Reply:
That's because a slot is freed within the executor thread, and on heavy load on a 32 cores machine, the slot may be reused before the thread is back to the pool.

Reply:
Oh, for BufferedMutator we are using SynchronousQueue...

Reply:
What's next? Is there something blocking the review?